### ISSUE TITLE
[BUG]: 

### ISSUE THREAD

**Issue by goffinf** [octokit/rest.js#331]
### What happened?

Attempting to use the conditions object when creating either an Organisation ruleset or a Repository ruleset always fails and returns a 422 status code.

If I remove the conditions structure the code creates the ruleset successfully

### Versions

Octokit rest.is version 20.0.1
NodeJS: latest

### Relevant log output

```shell
// Code example

import { Octokit } from "@octokit/rest";
import { RequestError } from "@octokit/request-error";
 
const octokit = new Octokit({
    auth: process.env.GITHUB_TOKEN
  });

/*
Create Org Rulest
*/

async function createOrgRuleset() {
    const org_ruleset = await octokit.rest.repos.createOrgRuleset({
        org: 'my-org',
        name: 'octo-ghas-test5',
        enforcement: 'evaluate',
        conditions: {
            ref_name: {
                include: [
                    '~ALL'
                ],
                exclude: [
                    'refs/head/dev*'
                ]
            },
            repository_name: {
                include: [
                    'octo-test'
                ]
            }
        },
        rules: [
            {
                type: 'required_status_checks',
                parameters: {
                    required_status_checks: [
                        {
                            context: 'CodeQL'
                        }
                    ],
                    strict_required_status_checks_policy: true
                }
            },
            {
                type: 'pull_request',
                parameters: {
                    require_code_owner_review: true,
                    require_last_push_approval: true,
                    required_approving_review_count: 2,
                    dismiss_stale_reviews_on_push: true,
                    required_review_thread_resolution: true
                }
            }
        ],
        bypass_actors: [
            {
                actor_id: 1,
                actor_type: 'OrganizationAdmin',
                bypass_mode: 'always'
            }
        ]
    });

    return org_ruleset;
};

try {
    console.log("Creating a Org ruleset");
    const org_ruleset = await createOrgRuleset();
    console.log("Org ruleset created successfully");
    console.log(org_ruleset);
    console.log(repo_ruleset);
} catch (error) {
    console.log("BOOOOOM !!!");
    console.log("Creating an Org Ruleset failed. The response status code was " + error.status)
} finally {
    console.log("Cleaning up");
}
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by goffinfnbs** [comment]
Some further detail.

As a comparison I attempt to create and organisation ruleset using CURL to see if the issue is specific to the Octokit REST client.

TLDR; It would appear not although the reason remain unclear.

I took the example directly from the API docs [here](https://docs.github.com/en/rest/orgs/rules?apiVersion=2022-11-28#create-an-organization-repository-ruleset), which incidentally also contains a JavaScipt example which also **includes** the **conditions** construct:.

But first the CURL command, adjusted to my Org and repo:

`curl -L -X POST -H "Accept: application/vnd.github+json" -H "Authorization: Bearer xxxxx" -H "X-GitHub-Api-Version: 2022-11-28" https://api.github.com/orgs/my-org/rulesets -d '{"name":"super cool ruleset9","target":"branch","enforcement":"evaluate","bypass_actors":[{"actor_id":1,"actor_type":"OrganizationAdmin","bypass_mode":"always"}],"conditions":{"ref_name":{"include":["refs/heads/main"],"exclude":["refs/heads/dev*"]},"repository_name":["octo-test"]},"rules":[{"type":"commit_author_email_pattern","parameters":{"operator":"contains","pattern":"github"}}]}'`

This was the response:

```
{
  "message": "Invalid request.\n\nInvalid property /conditions: data matches no possible input. See `documentation_url`.",
  "documentation_url": "https://docs.github.com/rest/orgs/rules#create-organization-repository-ruleset"
}
```
This would suggest **conditions** is **not supported** although going the the docs at the URL supplied we can see that it is described in detail and both the CULR and JavaScript examples include it ?

![image](https://github.com/octokit/rest.js/assets/126687169/eab4fd37-4cd3-4349-bc39-f04ac7b90593)

Interestingly I tried creating a ruleset at the **repository** level using CULR and was successful

`curl -L -X POST -H "Accept: application/vnd.github+json" -H "Authorization: Bearer xxxxxx" -H "X-GitHub-Api-Version: 2022-11-28" https://api.github.com/repos/my-org/octo-test/rulesets -d '{"name":"super cool ruleset5","target":"branch","enforcement":"evaluate","bypass_actors":[{"actor_id":1,"actor_type":"OrganizationAdmin","bypass_mode":"always"}],"conditions":{"ref_name":{"include":["refs/heads/main"],"exclude":["refs/heads/dev*"]}},"rules":[{"type":"commit_author_email_pattern","parameters":{"operator":"contains","pattern":"github"}}]}'`

with a response of:

```
{
  "id": 53370,
  "name": "super cool ruleset5",
  "target": "branch",
  "source_type": "Repository",
  "source": "my-org/octo-test",
  "enforcement": "evaluate",
  "conditions": {
    "ref_name": {
      "exclude": [
        "refs/heads/dev*"
      ],
      "include": [
        "refs/heads/main"
      ]
    }
  },
  "rules": [
    {
      "type": "commit_author_email_pattern",
      "parameters": {
        "name": "",
        "negate": false,
        "pattern": "github",
        "operator": "contains"
      }
    }
  ],
  "node_id": "RRS_lACqUmVwb3NpdG9yec4nfaTtzdB6",
  "_links": {
    "self": {
      "href": "https://api.github.com/repos/my-org/octo-test/rulesets/53370"
    },
    "html": {
      "href": "https://github.com/my-org/octo-test/rules/53370"
    }
  },
  "created_at": "2023-07-21T12:44:59.728+01:00",
  "updated_at": "2023-07-21T12:44:59.743+01:00",
  "bypass_actors": [
    {
      "actor_id": 1,
      "actor_type": "OrganizationAdmin",
      "bypass_mode": "always"
    }
  ],
  "current_user_can_bypass": "always"
}
```

This does **NOT work** with the Octokit client though

```
async function createRepoRuleset() {
    const org_ruleset = await octokit.rest.repos.createRepoRuleset({
        owner: 'my-org',
        repo: 'octo-test',
        name: 'octp-repo-ruleset7',
        enforcement: 'evaluate',
        target: 'branch',
        conditions: {
            ref_name: {
                include: [
                    'refs/heads/main'
                ],
                exclude: [
                    'refs/head/dev*'
                ]
            }
        },
        rules: [
            {
                type: 'required_status_checks',
                parameters: {
                    required_status_checks: [
                        {
                            context: 'CodeQL'
                        }
                    ],
                    strict_required_status_checks_policy: true
                }
            },
            {
                type: 'pull_request',
                parameters: {
                    require_code_owner_review: true,
                    require_last_push_approval: true,
                    required_approving_review_count: 2,
                    dismiss_stale_reviews_on_push: true,
                    required_review_thread_resolution: true
                }
            }
        ],
        bypass_actors: [
            {
                actor_id: 1,
                actor_type: 'OrganizationAdmin',
                bypass_mode: 'always'
            }
        ]
    });
    return org_ruleset;
};
```
which output:

```
request {
  method: 'POST',
  baseUrl: 'https://api.github.com',
  headers: {
    accept: 'application/vnd.github.v3+json',
    'user-agent': 'octokit-rest.js/20.0.1 octokit-core.js/5.0.0 Node.js/20.3.1 (linux; x64)'
  },
  mediaType: { format: '', previews: [] },
  request: { hook: [Function: bound bound register] },
  url: '/repos/{owner}/{repo}/rulesets',
  owner: 'my-org',
  repo: 'octo-test',
  name: 'octp-repo-ruleset7',
  enforcement: 'evaluate',
  target: 'branch',
  conditions: { ref_name: { include: [Array], exclude: [Array] } },
  rules: [
    { type: 'required_status_checks', parameters: [Object] },
    { type: 'pull_request', parameters: [Object] }
  ],
  bypass_actors: [
    {
      actor_id: 1,
      actor_type: 'OrganizationAdmin',
      bypass_mode: 'always'
    }
  ]
}
POST /repos/my-org/octo-test/rulesets - 422 in 459ms
```

**Comment by goffinfnbs** [comment]
Same behaviours with the GH CLI - Org rulesets FAIL, Repo rulesets SUCCEED.

So either I'm not doing something correctly with the Org level request payload or it genuinely is not supported ?

ORG level:

```
jq -n '{"name":"super cool ruleset9","target":"branch","enforcement":"evaluate","bypass_actors":[{"actor_id":1,"actor_type":"OrganizationAdmin","bypass_mode":"always"}],"conditions":{"ref_name":{"include":["refs/heads/main"],"exclude":["refs/heads/dev*"]},"repository_name":["cicd-ghas"]},"rules":[{"type":"commit_author_email_pattern","parameters":{"operator":"contains","pattern":"github"}}]}' | gh api --method POST -H "Accept: application/vnd.github+json" -H "X-GitHub-Api-Version: 2022-11-28" /orgs/my-org/rulesets --input -

{
  "message": "Invalid request.\n\nInvalid property /conditions: data matches no possible input. See `documentation_url`.",
  "documentation_url": "https://docs.github.com/rest/orgs/rules#create-organization-repository-ruleset"
}
gh: Invalid request.

Invalid property /conditions: data matches no possible input. See `documentation_url`. (HTTP 422)
```

REPO level:

```
jq -n '{"name":"super cool ruleset9","target":"branch","enforcement":"evaluate","bypass_actors":[{"actor_id":1,"actor_type":"OrganizationAdmin","bypass_mode":"always"}],"conditions":{"ref_name":{"include":["refs/heads/main"],"exclude":["refs/heads/dev*"]}},"rules":[{"type":"commit_author_email_pattern","parameters":{"operator":"contains","pattern":"github"}}]}' | gh api --method POST -H "Accept: application/vnd.github+json" -H "X-GitHub-Api-Version: 2022-11-28" /repos/my-org/octo-test/rulesets --input -


{
  "id": 53577,
  "name": "super cool ruleset9",
  "target": "branch",
  "source_type": "Repository",
  "source": "cicd-ghas/ghas-test",
  "enforcement": "evaluate",
  "conditions": {
    "ref_name": {
      "exclude": [
        "refs/heads/dev*"
      ],
      "include": [
        "refs/heads/main"
      ]
    }
  },
  "rules": [
    {
      "type": "commit_author_email_pattern",
      "parameters": {
        "name": "",
        "negate": false,
        "pattern": "github",
        "operator": "contains"
      }
    }
  ],
  "node_id": "RRS_lACqUmVwb3NpdG9yec4nfaTtzdFJ",
  "_links": {
    "self": {
      "href": "https://api.github.com/repos/cicd-ghas/ghas-test/rulesets/53577"
    },
    "html": {
      "href": "https://github.com/cicd-ghas/ghas-test/rules/53577"
    }
  },
  "created_at": "2023-07-21T16:09:01.287+01:00",
  "updated_at": "2023-07-21T16:09:01.316+01:00",
  "bypass_actors": [
    {
      "actor_id": 1,
      "actor_type": "OrganizationAdmin",
      "bypass_mode": "always"
    }
  ],
  "current_user_can_bypass": "always"
}
```

**Comment by jschnapper** [comment]
Hey there, the `repository_name` field properties are the same as the `ref_name` properties. So rather than this

```
    "conditions": {
        "repository_name": ["cicd-ghas"],
        },
        "ref_name": {
            "exclude": [
                "refs/heads/dev*"
            ],
            "include": [
                "refs/heads/main"
            ]
        }
    },
```

You would likely want to use

```
    "conditions": {
        "repository_name": {
            "exclude": [],
            "include": [
                "cicd-ghas"
            ]
        },
        "ref_name": {
            "exclude": [
                "refs/heads/dev*"
            ],
            "include": [
                "refs/heads/main"
            ]
        }
    },
```

And you can throw in `"protected": true` into the `repository_name` object to ensure the repo name can't be renamed, if desired.

You can see the expanded properties in the [docs](https://docs.github.com/en/rest/orgs/rules?apiVersion=2022-11-28#create-an-organization-repository-ruleset)

![Screenshot 2023-07-21 at 1 51 52 PM](https://github.com/octokit/rest.js/assets/15844781/3b991df7-e8f8-4914-ba6c-aea8c8576aad)

**Comment by goffinfnbs** [comment]
Thanks @jschnapper after a bit more tinkering that did resolve the problem in all 3 clients. Awesome that you took the time to explain so clearly and not just RTFM (even though it is clear that I didn't do that well enough).

I did note that you must include both an include AND exclude (even if its an empty array) for both the repository_name and ref_name.  The repository_name protected property is optional. That **is** kinda implied in the docs but did trip me up for a little while (I omitted empty arrays and kept seeing a 422).

All good now. Closing.

**Comment by goffinfnbs** [comment]
Happy to CLOSE

**Comment by goffinf** [comment]
Closed
